### PR TITLE
Light block: introduce useBlockWrapperProps

### PR DESCRIPTION
**Pull request by ellatrix**
## Description

This PR introduces `useBlockWrapperProps`, a hook used to lightly mark any element as a block.

To use, just pass the returned props to the element:

```jsx
<p attribute="value" { ...useBlockWrapperProps() } />
```

If the component defines a ref for the element, it must be passed through `useBlockWrapperProps`. If no ref is defined, `useBlockWrapperProps` will create one and pass it to the element.

```jsx
<p attribute="value" { ...useBlockWrapperProps( { ref } ) } />
```

You can pass any props to `useBlockWrapperProps` and they will be returned, but you don't _have_ to. In case of `style` and `className` props, `useBlockWrapperProps` will automatically merge them. If you don't pass them through `useBlockWrapperProps`, you have to merge them yourself.

Both are equally ok:

```jsx
<p { ...useBlockWrapperProps( { ref, className: 'something', attribute: value } ) } />
```

```jsx
const blockProps = useBlockWrapperProps( { ref } );
<p attribute="value" { ... blockProps } className={ classnames( blockProps.className, 'something' ) } />
```

A hook means that it can work for any component. No need to have a `<Block as />` component and `block()` HoC so that it works in different scenarios. Just one API. We don't need access to the element/component at all, we only need to define some props and have access to the ref.

I think it's really nice that this leaves the original element structure intact, since it doesn't need to be wrapped in another component.

I could imagine a similar API for `InnerBlocks`, which would avoid having things like the ugly `passedProps` etc.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

**Comment by ZebulanStanphill** [comment]
Wow, this looks really neat! I definitely prefer this to the previous approaches.

One question: should we call it `useBlockProps` or `useBlockWrapperProps`?

**Comment by ellatrix** [comment]
I don't mind either way... I slightly prefer something shorter, but at the end of the day it doesn't matter much.

File: packages/block-editor/src/components/block-list/block-wrapper.js [hunk 1]
```diff
@@ -185,37 +191,82 @@ const BlockComponent = forwardRef(
...
+const BlockComponent = forwardRef(
+	(
+		{ children, tagName: TagName = 'div', __unstableIsHtml, ...props },
+		wrapper
+	) => {
+		const wrapperProps = useBlockProps(
+			{ ref: wrapper },
+			{ __unstableIsHtml }
+		);
+		const isAligned = wrapperProps && !! wrapperProps[ 'data-align' ];
```
Comment (youknowriad) [inline packages/block-editor/src/components/block-list/block-wrapper.js:255]: There's still a lot of things remaining here. alignment support, labels... role. Can't we absorb everything?

File: packages/block-editor/src/components/block-list/block.js [hunk 2]
```diff
@@ -187,7 +188,7 @@ function BlockListBlock( {
 		name,
 		mode,
 		blockTitle: blockType.title,
-		wrapperProps,
+		wrapperProps: omit( wrapperProps, [ 'data-align' ] ),
```
Comment (youknowriad) [inline packages/block-editor/src/components/block-list/block.js:191]: Not for this PR but that whole object "value" is mostly important for the wrapper hook and not for the current component.

File: packages/block-editor/src/components/block-list/block-wrapper.js [hunk 3]
```diff
@@ -184,43 +192,69 @@ const BlockComponent = forwardRef(
...
+			'wp-block',
+			className,
+			props.className,
+			wrapperProps.className
+		),
+		style: { ...wrapperProps.style, ...props.style },
+	};
+}
+
+const BlockComponent = forwardRef(
```
Comment (youknowriad) [inline packages/block-editor/src/components/block-list/block-wrapper.js:247]: This is cool, when do you think we should remove this? should we add a deprecated call with a version?

**Review by youknowriad — COMMENTED** [review]
I was wondering if there's a way to make this even simpler for block authors. Instead of telling them to apply a hook, we could pass the "ref" as a prop and all they'll have to do would be to apply the ref? how possible is that?

File: packages/block-editor/src/components/block-list/block.js (hunk 2 above)
Comment (ellatrix) [reply packages/block-editor/src/components/block-list/block.js:191]: Yes, that's true. The memoized value as context should ideally be reduced or even avoided entirely.

**Comment by ellatrix** [comment]
@youknowriad That's an excellent idea! I'll explore it quick.

File: packages/block-editor/src/components/block-list/block-wrapper.js (hunk 3 above)
Comment (ellatrix) [reply packages/block-editor/src/components/block-list/block-wrapper.js:247]: Since it's experimental, maybe we can just remove it. Depending on what we do, we might have to leave it since some block components are still classes. Maybe the ref idea could replace it though.

**Comment by ellatrix** [comment]
@youknowriad Oh, actually, I don't think it would be that easy because we still have to deal with the other props that need to be passed down to the block. :/

**Comment by youknowriad** [comment]
> @youknowriad Oh, actually, I don't think it would be that easy because we still have to deal with the other props that need to be passed down to the block. :/

One possibility would be to apply the style and classname using DOM apis and not React props but not sure how good is that

File: packages/block-editor/src/components/block-list/block-wrapper.js (hunk 3 above)
Comment (youknowriad) [reply packages/block-editor/src/components/block-list/block-wrapper.js:247]: We could yes, I saw some third-party usage already. So let's be nice to them and deprecate for a release or two.

**Comment by ellatrix** [comment]
@youknowriad I was thinking about that, and it would work fine if there were a limited amount of props controlled by us, but we have a filter on the block edit component where plugins are able to add additional props, so I think it's better to pass them straight to the react element.

**Comment by youknowriad** [comment]
Starting to wonder whether the hook is the best API. I mean it is but with third-party random props we can already have issues as blocks using the hook won't "merge" all props (imagine blocks having attached event handlers).

I'd love if we can officially restrict that hook and make it more predictable, it's harming us in a lot of ways.

**Comment by ellatrix** [comment]
> imagine blocks having attached event handlers

Event handlers are ok. We don't attach any event handlers with react.

> I'd love if we can officially restrict that hook and make it more predictable, it's harming us in a lot of ways.

What do you mean?

**Comment by ellatrix** [comment]
Let me elaborate :)

1. We don't need access to the rendered element at all, only to the attributes/props. There's no need to concern ourselves with what element or component the block renders.
2. It's true that some props need to be merged. This will be the case with a `block()` HoC and `<Block as={ Component } />` too.
3. To reduce the chance of conflicts, I've moved our event listeners to be attached in effects. React only allows one event listener, while the DOM allows many, so we can leave the React slot open for use by the block.
4. Of course there's still a change of conflict, but that's always been the case. If both a block and a block edit filter (or multiple block edit filters by multiple plugins) add `onClick` props, only one will make it. This has always been a problem.

**Comment by gziolo** [comment]
Awesome work @ellatrix, I like where it is heading. The proposed API is way much better than the `Block.*` the notation used previously 💯 

I agree with @ZebulanStanphill that `useBlockWrapperProps` would be less confusing for those who see it for the first time. I think it's very important to name it properly so it would be clear when it should be used. Does it also mean that with this hook we could somehow get rid of the supports flag for the light wrapper?

**Comment by youknowriad** [comment]
> What do you mean?

I mean can we add a warning to the console when we detect that there are unsupported wrapper props or props (all what's not className and style for me)

**Comment by ellatrix** [comment]
Ok, let's make it `useBlockWrapperProps` then. :)

No, the supports flag will have to stay I think... I don't think there's any other way for the parent to know otherwise if it should render a block wrapper in case the block itself doesn't render one.

**Comment by ellatrix** [comment]
> I mean can we add a warning to the console when we detect that there are unsupported wrapper props or props (all what's not className and style for me)

Oh, from the block edit filter? Sure, I'd like that :)

**Comment by ellatrix** [comment]
I even start getting confused myself if we're talking about React hooks and WP hooks at the same time... :)

File: packages/block-library/src/image/edit.js [hunk 4]
```diff
@@ -275,7 +275,7 @@ export function ImageEdit( {
 	return (
 		<>
 			{ controls }
-			<Block.figure ref={ ref } className={ classes } key={ key }>
+			<figure { ...useBlockProps( { key, ref, className: classes } ) }>
```
Comment (gziolo) [inline packages/block-library/src/image/edit.js:278]: I think it's fine to leave `key` as it was, it's used only to help React with re-renders.

File: packages/block-library/src/list/edit.js [hunk 5]
```diff
@@ -178,6 +178,7 @@ export default function ListEdit( {
 				start={ start }
 				reversed={ reversed }
 				type={ type }
+				{ ...useBlockProps() }
```
Comment (gziolo) [inline packages/block-library/src/list/edit.js:181]: `useBlockProps` is a React hook, shouldn't it be at the top of the component here and it other places to ensure it is always predictable (according to [rules of hooks](https://en.reactjs.org/docs/hooks-rules.html))?

Shouldn't those spread props be applied as the first item? In the current shape, if there is any clash in naming, block wrapper props will erase previously listed props.

File: packages/block-editor/src/components/block-list/block-wrapper.js [hunk 6]
```diff
@@ -28,163 +28,176 @@ import { Context, SetBlockNodes } from './root-container';
...
+			if ( target !== ref.current || isTextField( target ) ) {
 				return;
 			}
 
 			event.preventDefault();
 
-			if ( keyCode === ENTER ) {
-				insertDefaultBlock( {}, rootClientId, index + 1 );
-			} else {
-				removeBlock( clientId );
```
Comment (youknowriad) [inline packages/block-editor/src/components/block-list/block-wrapper.js:183]: Why did you remove the "removal" handling? is this handled elsewhere?

Comment (ellatrix) [reply packages/block-editor/src/components/block-list/block-wrapper.js:183]: There's a shortcut for it defined as well.

File: packages/block-library/src/list/edit.js (hunk 5 above)
Comment (ellatrix) [reply packages/block-library/src/list/edit.js:181]: I think it's fine to inline if possible. We should also try to avoid early returns in components if hooks are being used. It's pointless to execute the hook otherwise.

**Comment by ellatrix** [comment]
Not sure why the native list block is failing. I've exported `useBlockProps` in the native file as well. Do you have any idea, @SergioEstevao?

File: packages/block-library/src/image/edit.js (hunk 4 above)
Comment (ellatrix) [reply packages/block-library/src/image/edit.js]: I've removed it now. We should avoid triggering a remount.

File: packages/block-editor/src/components/block-list/block-wrapper.js (hunk 6 above)
Comment (youknowriad) [reply packages/block-editor/src/components/block-list/block-wrapper.js:183]: I guess we should adapt the comment and should the "Enter" be handled with a shortcut for consistency?

File: packages/block-library/src/list/edit.js (hunk 5 above)
Comment (talldan) [reply packages/block-library/src/list/edit.js:181]: I'm with @gziolo here. I don't think hooks should be inlined in this way. It's fine technically, but it seems like it'd be very easy for devs that might be changing this code later to miss the presence of this hook.

File: packages/block-library/src/audio/edit.js [hunk 7]
```diff
@@ -116,7 +116,7 @@ function AudioEdit( {
 	}
 	if ( ! src ) {
 		return (
-			<Block.div>
+			<div { ...blockProps() }>
```
Comment (ZebulanStanphill) [inline packages/block-library/src/audio/edit.js:119]: Misplaced `()` here:
```suggestion
			<div { ...blockProps }>
```

File: packages/block-library/src/audio/edit.js [hunk 8]
```diff
@@ -175,7 +175,7 @@ function AudioEdit( {
 					/>
 				</PanelBody>
 			</InspectorControls>
-			<Block.figure>
+			<figure { ...blockProps() }>
```
Comment (ZebulanStanphill) [inline packages/block-library/src/audio/edit.js:178]:
```suggestion
			<figure { ...blockProps }>
```

File: packages/block-library/src/button/edit.js [hunk 9]
```diff
@@ -171,7 +171,7 @@ function ButtonEdit( props ) {
 	return (
 		<>
 			<ColorEdit { ...props } />
-			<Block.div>
+			<div { ...useBlockProps() }>
```
Comment (ZebulanStanphill) [inline packages/block-library/src/button/edit.js:174]: For consistency, I think we should avoid inlining any of these hook calls. If an if statement was ever added to implement something like an early return, this would suddenly break the rules of hooks.

**Review by ZebulanStanphill — DISMISSED** [review]
I found some typos that _might_ be the cause of the test failures. Also, this needs a rebase.

File: packages/block-editor/src/components/index.native.js [hunk 10]
```diff
@@ -50,7 +50,10 @@ export { default as BlockToolbar } from './block-toolbar';
 export { default as DefaultBlockAppender } from './default-block-appender';
 export { default as __unstableEditorStyles } from './editor-styles';
 export { default as Inserter } from './inserter';
-export { Block as __experimentalBlock } from './block-list/block-wrapper';
+export {
+	Block as __experimentalBlock,
+	useBlockProps as __experimentalUseBlockProps,
+} from './block-list/block-wrapper';
```
Comment (etoledom) [inline packages/block-editor/src/components/index.native.js:56]: Thank you @ellatrix for updating the `native.js` side of this!

It seems that `useBlockProps` was defined in `block-list/block-wrapper.js `, but we have a `block-wrapper.native.js` version of `block-wrapper` without `useBlockProps`. 

Without going deep into the native errors, this seems to be at least part of the problem.

**Comment by etoledom** [comment]
I've done some tests on Mobile.

It was possible to reproduce the error on the native Unit Tests loading a Quote block and a List block.




Apart from that, the Alignment feature seems to be broken, not working in Paragraph, Heading, Image Captions, etc...

File: packages/block-editor/src/components/rich-text/index.js [hunk 11]
```diff
@@ -526,7 +530,6 @@ function RichTextWrapper(
 
 	const content = (
 		<RichText
-			{ ...props }
```
Comment (etoledom) [inline packages/block-editor/src/components/rich-text/index.js:529]: This seems to by what is breaking Alignment on Mobile.

By removing ` { ...props } ` we are missing some important Mobile props:

These are some of those printed con console. Not sure if there might be more:
Most of this is text styling information that we can not pass via CSS.

```js
deleteEnter: true
fontSize: 12
onBlur: undefined
placeholderTextColor: "#ccc"
rootTagsToEliminate: ["p"]
textAlign: "center".  // ---> Alignment info that was missing
underlineColorAndroid: "transparent"
```

**Comment by etoledom** [comment]
> It was possible to reproduce the error on the native Unit Tests loading a Quote block and a List block.

This crash is due to what was explained [here](https://github.com/WordPress/gutenberg/pull/23034#discussion_r464295571)

List and Quote are have shared `edit.js` files between platforms. But `Block.js` and `Block-wrapper` have a `.native.js` variant.

I tried a dumb `useBlockProps` definition in `block-wrapper.native.js`:

```js
export function useBlockProps( props = {}, { __unstableIsHtml } = {} ) {
	return {
		...props,
		style: {...props.style },
	};
}
```

Not sure if there would be implications or if we are missing important props on mobile for Quote and List, but a fast smoke test seems to work fine (together with adding back `{ ...props }` [here](https://github.com/WordPress/gutenberg/pull/23034#discussion_r464356825))

To go deeper and implement this properly, I believe we would need to refactor `block.native.js` as a function and implement `BlockListBlockContext` following `block.js`. But again, not sure if this is needed if the dumb implementation is enough.

cc @hypest

File: packages/block-library/src/list/edit.js (hunk 5 above)
Comment (noisysocks) [reply packages/block-library/src/list/edit.js:181]: We have the `react-hooks/rules-of-hooks` ESLint rule configured, no? That should catch any mistakes due to using the hook in a branch.

Comment (noisysocks) [reply packages/block-library/src/list/edit.js:181]: Yep, I tested locally and `react-hooks/rule-of-hooks` will complain if you return before using `useBlockProps`.

```
gutenberg (try/use-block *) % git diff
diff --git a/packages/block-library/src/list/edit.js b/packages/block-library/src/list/edit.js
index a5094cb569..e283bc3b3a 100644
--- a/packages/block-library/src/list/edit.js
+++ b/packages/block-library/src/list/edit.js
@@ -51,6 +51,10 @@ export default function ListEdit( {
                return !! select( 'core/block-editor' ).getSettings().isRTL;
        }, [] );
 
+       if ( isRTL ) {
+               return null;
+       }
+
        const controls = ( { value, onChange, onFocus } ) => (
                <>
                        { isSelected && (
gutenberg (try/use-block *) % npm run lint-js packages/block-library/src/list/edit.js

> gutenberg@8.4.0-rc.1 lint-js /Users/robert/Code/gutenberg
> wp-scripts lint-js "packages/block-library/src/list/edit.js"


/Users/robert/Code/gutenberg/packages/block-library/src/list/edit.js
  185:10  error  React Hook "useBlockProps" is called conditionally. React Hooks must be called in the exact same order in every component render  react-hooks/rules-of-hooks

✖ 1 problem (1 error, 0 warnings)

npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! gutenberg@8.4.0-rc.1 lint-js: `wp-scripts lint-js "packages/block-library/src/list/edit.js"`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the gutenberg@8.4.0-rc.1 lint-js script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/robert/.npm/_logs/2020-08-04T02_32_25_789Z-debug.log
```

File: packages/block-editor/src/components/rich-text/index.js (hunk 11 above)
Comment (ellatrix) [reply packages/block-editor/src/components/rich-text/index.js:531]: @etoledom What component should be receiving these props? The internal RichText component or the wrapper element itself?

Comment (ellatrix) [reply packages/block-editor/src/components/rich-text/index.js:531]: Cc @SergioEstevao

Comment (ellatrix) [reply packages/block-editor/src/components/rich-text/index.js:531]: Ok, figured it out, I added the props.

Comment (hypest) [reply packages/block-editor/src/components/rich-text/index.js:531]: > Cc SergioEstevao

Thanks for reeling people in Ella! Just a quick update that Sérgio is on his 3-months sabbatical and it's best to instead ping @etoledom (iOS), or @cameronvoell (Android) or @guarani (iOS) on RichText matters. Thanks!

File: packages/block-library/src/navigation/edit.js [hunk 12]
```diff
@@ -55,6 +55,8 @@ function Navigation( {
 
 	const { selectBlock } = useDispatch( 'core/block-editor' );
 
+	const blockProps = useBlockWrapperProps( { ref } );
```
Comment (ZebulanStanphill) [inline packages/block-library/src/navigation/edit.js:58]: Does this ref need to be passed to the hook? It's not placed on the block's root element, but rather the contained `<NavigationPlaceholder />`.

Comment (ellatrix) [reply packages/block-library/src/navigation/edit.js:58]: Oh, you're right. Maybe a rebase error. :)

File: packages/block-editor/src/components/block-list/block-wrapper.js [hunk 13]
```diff
@@ -28,148 +28,160 @@ import { Context, SetBlockNodes } from './root-container';
...
+		// Find all tabbables within node.
+		const textInputs = focus.tabbable
+			.find( ref.current )
+			.filter( isTextField )
+			// Exclude inner blocks and block appenders
+			.filter(
+				( node ) =>
+					isInsideRootBlock( ref.current, node ) &&
+					! node.closest( '.block-list-appender' )
+			);
```
Comment (ZebulanStanphill) [inline packages/block-editor/src/components/block-list/block-wrapper.js:142]: Not strictly related to this PR, but we could optimize this by only filtering once.
```suggestion
			.filter(
				( node ) =>
					isTextField( node ) &&
					// Exclude inner blocks and block appenders.
					isInsideRootBlock( ref.current, node ) &&
					! node.closest( '.block-list-appender' )
			);
```

File: packages/block-library/src/image/image.js [hunk 14]
```diff
@@ -80,6 +80,8 @@ export default function Image( {
 	onUploadError,
 	containerRef,
 } ) {
+	const captionRef = useRef();
+	const prevUrl = useRef();
```
Comment (ZebulanStanphill) [inline packages/block-library/src/image/image.js:84]: There's a dedicated hook for this sort of thing in the `compose` package: [`usePrevious`](https://github.com/WordPress/gutenberg/blob/master/packages/compose/src/hooks/use-previous/README.md).

File: packages/block-editor/src/components/block-list/block-wrapper.js [hunk 15]
```diff
@@ -28,148 +28,160 @@ import { Context, SetBlockNodes } from './root-container';
...
+ * hook and pass the returned props to the element to mark as a block. If you
+ * define a ref for the element, it is important to pass the ref to this hook,
+ * which the hooks in turn will pass to the component through the props it
+ * returns. Optionally, you can also pass any other props through this hook, and
+ * they will be merged and returned.
+ *
+ * @param {Object}  props   Optional. Props to pass to the element. Must contain
+ *                          the ref if one is defined.
+ * @param {Object}  options Options for internal use only.
+ * @param {boolean} options.__unstableIsHtml
```
Comment (ZebulanStanphill) [inline packages/block-editor/src/components/block-list/block-wrapper.js:42]: Why put the unstable props in a separate parameter? They're already labeled with the `__unstable` prefix.

Comment (ellatrix) [reply packages/block-editor/src/components/block-list/block-wrapper.js:42]: The linter logs a warning if it's not documented.

File: packages/block-editor/src/components/rich-text/index.js [hunk 16]
```diff
@@ -128,6 +128,22 @@ function RichTextWrapper(
...
+		textAlign,
+		selectionColor,
+		tagsToEliminate,
+		rootTagsToEliminate,
+		disableEditingMenu,
+		fontSize,
+		fontFamily,
+		fontWeight,
+		fontStyle,
 		...props
```
Comment (youknowriad) [inline packages/block-editor/src/components/rich-text/index.js:147]: Can you clarify the changes to this component more? I mean why the props need to end up on different components now?

File: packages/block-library/src/button/edit.js [hunk 17]
```diff
@@ -201,7 +201,7 @@ function ButtonEdit( props ) {
 	return (
 		<>
 			<ColorEdit { ...props } />
-			<Block.div>
+			<div { ...useBlockWrapperProps() }>
```
Comment (youknowriad) [inline packages/block-library/src/button/edit.js:204]: I think we should set a guideline that we'd avoid to use this inline to avoid the React hooks order issue even if the component doesn't have early returns (they can be added inadvertently later).

File: packages/block-library/src/image/image.js [hunk 18]
```diff
@@ -142,6 +144,16 @@ export default function Image( {
...
 
+	// Focus the caption after inserting an image from the placeholder. This is
+	// done to preserve the behaviour of focussing the first tabbable element
+	// when a block is mounted. Previously, the image block would remount when
+	// the placeholder is removed. Maybe this behaviour could be removed.
+	useEffect( () => {
+		if ( url && ! prevUrl && isSelected ) {
+			captionRef.current.focus();
+		}
+	}, [ url, prevUrl ] );
```
Comment (youknowriad) [inline packages/block-library/src/image/image.js:155]: Why we need to introduce this particular behavior in this PR? Is there a different between the previous implementation of Block.div and useBlockEditWrapper?

File: packages/e2e-tests/specs/editor/various/toolbar-roving-tabindex.test.js [hunk 19]
```diff
@@ -65,17 +65,17 @@ describe( 'Toolbar roving tabindex', () => {
...
 	} );
 
 	it( 'ensures list block toolbar uses roving tabindex', async () => {
 		await insertBlock( 'List' );
 		await page.keyboard.type( 'List' );
-		await testBlockToolbarKeyboardNavigation( 'Write list…' );
+		await testBlockToolbarKeyboardNavigation( 'Block: List' );
 		await wrapCurrentBlockWithGroup();
-		await testGroupKeyboardNavigation( 'Write list…' );
+		await testGroupKeyboardNavigation( 'Block: List' );
```
Comment (youknowriad) [inline packages/e2e-tests/specs/editor/various/toolbar-roving-tabindex.test.js:78]: What changes that lead to this?

Comment (ellatrix) [reply packages/e2e-tests/specs/editor/various/toolbar-roving-tabindex.test.js:78]: Change in prop priority. The label should be assigned by us (the editor), not the block.

File: packages/block-library/src/button/edit.js (hunk 17 above)
Comment (ellatrix) [reply packages/block-library/src/button/edit.js:204]: When would there be an order issue when there is not early returns?

Comment (ellatrix) [reply packages/block-library/src/button/edit.js:204]: But I'm fine with avoiding inline uses. I don't have a strong opinion.

File: packages/block-editor/src/components/rich-text/index.js (hunk 16 above)
Comment (ellatrix) [reply packages/block-editor/src/components/rich-text/index.js:147]: Because we may be spreading more props onto RichText now. Previously we cherry picked the props that should be passed down to the editable element such as style and classes. Now any prop that is not RichText's props will be passed down.

File: packages/block-library/src/image/image.js (hunk 18 above)
Comment (ellatrix) [reply packages/block-library/src/image/image.js:155]: Previously we were refocussing the block using `key`. Can be done separately I guess.

https://github.com/WordPress/gutenberg/pull/23034/files/808b311228c7da55f7e4f7552133fc0be26f97c8#diff-f5fc3d5a78f58a5274bc196ebc657c5bL275

Comment (ellatrix) [reply packages/block-library/src/image/image.js:155]: Split in https://github.com/WordPress/gutenberg/pull/25493

File: packages/block-library/src/button/edit.js (hunk 17 above)
Comment (youknowriad) [reply packages/block-library/src/button/edit.js:204]: > When would there be an order issue when there is not early returns?

never, it's just that someone editing the block in the future, might not even notice that there's a hook in the returned function. It's not common to use hooks in the returned element in React components.

**Comment by etoledom** [comment]
An update about https://github.com/WordPress/gutenberg/pull/23034#issuecomment-667976409
Everything seems to be working perfectly now on Mobile. 
Thank you for the update @ellatrix ! 🙏

File: packages/block-library/src/button/edit.js (hunk 17 above)
Comment (ellatrix) [reply packages/block-library/src/button/edit.js:204]: Sure, I'll adjust it :)

**Comment by youknowriad** [comment]
Can we add a deprecation for the `Block.*` components so we can remove it before landing on Core?

**Review by youknowriad — APPROVED** [review]

File: packages/block-library/src/button/edit.js (hunk 17 above)
Comment (ellatrix) [reply packages/block-library/src/button/edit.js]: Done

**Comment by ellatrix** [comment]
@youknowriad Yes, I'll add a deprecation.

File: packages/block-editor/src/components/block-list/block-wrapper.js (hunk 15 above)
Comment (ZebulanStanphill) [reply packages/block-editor/src/components/block-list/block-wrapper.js:42]: No, I mean why is it under a second `options` object, rather than the same one that contains the other parameters? Why is the function signature `useBlockWrapperProps( { ref }, { __unstableIsHtml } )` rather than `useBlockWrapperProps( { ref, __unstableIsHtml } )`?

Comment (youknowriad) [reply packages/block-editor/src/components/block-list/block-wrapper.js:42]: I believe because the first one is "props" to be merged with the output and the second one is settings for the hook itself.

Comment (ZebulanStanphill) [reply packages/block-editor/src/components/block-list/block-wrapper.js:42]: Oh, okay. I get it now. :+1:

**Comment by ZebulanStanphill** [comment]
Notably, the following blocks still use the `Block.TAG_NAME` API:
- Column
- Columns
- Group
- Heading
- Post Title
- Preformatted
- Site Tagline
- Site Title
- Social Icons
- Template Part
- Verse
- Video

**Comment by ellatrix** [comment]
@ZebulanStanphill Yes, remaining ones should be done in #25515.
